### PR TITLE
feat(ContentPresenter): Content of ContentPresenter should become DataContext of the subtree whenever ContentTemplate is not null

### LINKED ISSUE #6642: Content of ContentPresenter should become DataContext of the subtree whenever ContentTemplate is not null
Currently the only criteria for ContentPresenter.Content to become DataContext of the subtree is when it does not implement IControl:

    // Set the DataContext if the data isn't a control.
    if (!(content is IControl))
    {
        DataContext = content;
    }
    else
    {
        ClearValue(DataContextProperty);
    }

but in WPF - Content becomes DataContext also whenever ContentTemplate is non-null. I think WPF way is better, because many times you want the sub-tree defined by ContentTemplate to bind its property to some visual object that plays role of DataContext of the sub-tree.

### PR DESCRIPTION
**Pull request by workgroupengineering**
## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes #6642

### CODE REVIEW

**Comment by grokys** [comment]
Could we get some unit tests for this? `ContentPresenter` is pretty heavily tested at the moment, and doesn't make sense to not cover this is tests too.

File: src/Avalonia.Controls/Presenters/ContentPresenter.cs [hunk 1]
```diff
@@ -254,8 +266,14 @@ public sealed override void ApplyTemplate()
         public void UpdateChild()
         {
             var content = Content;
+            UpdateChid(content);
+        }
+
+        private void UpdateChid(object content)
```
Comment (jp2masa) [inline src/Avalonia.Controls/Presenters/ContentPresenter.cs:272]: Nit: typo

Comment (workgroupengineering) [reply src/Avalonia.Controls/Presenters/ContentPresenter.cs]: tnx

**Comment by workgroupengineering** [comment]
> Could we get some unit tests for this? `ContentPresenter` is pretty heavily tested at the moment, and doesn't make sense to not cover this is tests too.

I think there is no need for it, there are so many tests that verify it.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018395-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by jmacato** [comment]
Closing due to inactivity. Please ping me if you're gonna work on this again.

**Comment by workgroupengineering** [comment]
@jmacato This PR is complete. No further tests are needed because the behavior is extensively tested.

**Comment by jmacato** [comment]
@workgroupengineering okay i'll reopen and ask for reviews

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019181-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by jmacato** [comment]
@workgroupengineering please add additional test on this PR so that we dont break this in the future

**Review by maxkatz6 — APPROVED** [review]
LGTM

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019193-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: tests/Avalonia.Controls.UnitTests/Presenters/ContentPresenterTests_InTemplate.cs [hunk 2]
```diff
@@ -353,6 +353,19 @@ public void Should_Clear_Host_When_Host_Template_Cleared()
             Assert.Null(target.Host);
         }
 
+        [Fact]
+        public void Content_Will_By_Datacontext_When_ControlTemplate_Is_Null()
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/Presenters/ContentPresenterTests_InTemplate.cs:357]: `Content_Will_`?? Might be missing some word here?

**Review by jmacato — CHANGES_REQUESTED** [review]
just a nitpick :)

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019230-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by jmacato — APPROVED** [review]
LGTM!

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019436-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**